### PR TITLE
Implementation of Prefix domain

### PR DESCRIPTION
**Pull request by BlueRyse**
WIP

### CODE REVIEW

File: lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Prefix.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    @Override
+    protected Prefix wideningAux(Prefix other) throws SemanticException {
+        return lubAux(other);
+    }
+
+    @Override
+    protected boolean lessOrEqualAux(Prefix other) throws SemanticException {
+        return false;
```
Comment (VincenzoArceri) [inline lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Prefix.java]: We should implement this method.

File: lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Prefix.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    @Override
+    public Prefix top() {
+        return TOP;
+    }
+
+    @Override
+    public Prefix bottom() {
+        return BOTTOM;
+    }
+
```
Comment (VincenzoArceri) [inline lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Prefix.java]: We should override `isTop` and `isBottom` methods.

File: lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Prefix.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        if(constant.getValue() instanceof String)
+            return new Prefix( (String) constant.getValue() );
+
+        return TOP;
+    }
+
+    @Override
+    protected Prefix evalUnaryExpression(UnaryOperator operator, Prefix arg, ProgramPoint pp){
+        if(operator == StringLength.INSTANCE){
+            return this;
```
Comment (VincenzoArceri) [inline lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Prefix.java]: In this case we should return `TOP`

File: lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Prefix.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        else
+            return TOP;
+    }
+
+    @Override
+    protected Prefix evalBinaryExpression(BinaryOperator operator, Prefix left, Prefix right, ProgramPoint pp) throws SemanticException {
+        if(operator == StringConcat.INSTANCE){
```
Comment (VincenzoArceri) [inline lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Prefix.java]: For all the symbolic expressions for which we cannot return a prefix (e.g., StringContains), we should return `TOP`, not the least upper bound between `left` and `right`

**Review by VincenzoArceri — CHANGES_REQUESTED** [review]

**Comment by BlueRyse** [comment]
@VincenzoArceri Requested changes implemented
Ready for review

File: lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Prefix.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+        int i = 0;
+        while (i <= prefix.length() - 1 &&
+                i <= otherPrefixString.length() - 1 &&
+                prefix.charAt(i) == otherPrefixString.charAt(i)) {
+            result.append(prefix.charAt(i++));
+        }
+
+        if (result.length() != 0)
+            return new Prefix(result.toString());
```
Comment (VincenzoArceri) [inline lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Prefix.java]: ToString not needed here

**Review by VincenzoArceri — APPROVED** [review]

**Pull request merged**